### ISSUE TITLE
BINARY fields are truncated incorrectly

### ISSUE THREAD

**Issue by vicary** [sequelpro/sequelpro#2284]
Binary fields are automatically translated into base16 numbers `0x1234567890ABCDEF`, but when I copy and pasted into another BINARY field, it truncates the length as if it is some kind of string.

For example `BINARY(16)` returns a string with length 34 which starts with `"0x"` and followed by a HEX number with the length of 32 characters, but I can only enter 16 characters including `"0x"` which is only 7 bytes of data.

**Comment by dmoagx** [comment]
> Binary fields are automatically translated into base16 numbers
That is a setting you can make in the View menu. Hex field will be highlighted blue.

> but I can only enter 16 characters including "0x" which is only 7 bytes of data.
Sequel Pro does not allow editing in Hex mode at all.

This leads me to believe you are actually viewing the raw value of the field

**Comment by vicary** [comment]
Yes, it is Text mode from the Pop-up sheet.

If I turn off pop-up editing, I was able to kind of edit the field in raw scrambled characters. Not sure if it is about charset conversion, but copy and paste a value originally valid will results in a "data too long" error from MySQL.

**Comment by dmoagx** [comment]
I still have no idea what it is you are doing **exactly**

Also, are the hex values colored blue in the table view?

**Comment by vicary** [comment]
Nope, because with the "blue" option I am not allowed to edit the field in any way, I'll leave that alone.

Let me explain it with screenshots, the field is BINARY(16) of packed UUID value as shown.

![screen shot 2015-10-15 at 12 48 11 pm](https://cloud.githubusercontent.com/assets/85772/10505050/627b1888-733b-11e5-84c2-d9129cbd2861.png)

# With Pop-up sheet
![screen shot 2015-10-15 at 12 48 31 pm](https://cloud.githubusercontent.com/assets/85772/10505066/724c4c32-733b-11e5-8ff6-9527d42ccb9d.png)
This is how the value is displayed in pop-up sheet.

![screen shot 2015-10-15 at 12 48 44 pm](https://cloud.githubusercontent.com/assets/85772/10505067/72510fc4-733b-11e5-95f9-29430ef63a34.png)
When I copy & paste in the same box, it trims the string as if it is a VARCHAR(16), or simply string, whatever you may call it. The pop-up sheet seems unaware of the column type it is editing.

# Without Pop-up sheet
This is more easily explained by inserting instead of the update I was mentioning in the last comment.

![screen shot 2015-10-15 at 12 56 18 pm](https://cloud.githubusercontent.com/assets/85772/10505115/3f25dfac-733c-11e5-8fca-4bfae2eef6e0.png)
I copied the binary text from a field.

![screen shot 2015-10-15 at 12 56 34 pm](https://cloud.githubusercontent.com/assets/85772/10505117/3f308614-733c-11e5-9964-741d0bf2ba51.png)
Paste it in a new row.

![screen shot 2015-10-15 at 12 56 40 pm](https://cloud.githubusercontent.com/assets/85772/10505116/3f307ba6-733c-11e5-8b1e-16aceef04dcb.png)
And got this error.

# Conclusion

Because we have both the connection charset and the display charset, theoretically it can copy and paste the whole value without going raw SQL statements. Mac is able to copy invisible characters as well if they are within the selection range, which means only if there are invisible characters at the beginning or the end of the value, we shall then prone to data corruption.

**Comment by TheDigitalOrchard** [comment]
The MySQL documentation states the following:

> The permissible maximum length is the same for BINARY and VARBINARY as it is for CHAR and VARCHAR, except that the length for BINARY and VARBINARY is a length in bytes rather than in characters.

So it sounds like Sequel Pro is treating the Binary data as characters than than bytes when pasting. Wild guess on my part based on what I've read and seen here.

**Comment by vicary** [comment]
@TheDigitalOrchard Thanks for the info.

Took a glance in the code, this might be related to a default textbox component validation, which the max length cannot be skipped easily in the context of underlying field.

I guess as long as the database as it's own validation (see my **Without Popup sheet*** use case), the quickest way to fix this would be taking away the input validation altogether and let the database throw.

Because you won't know if the user is entering `0x` numbers or the scrambled text directly.

**Comment by dmoagx** [comment]
I feel like there might be an open issue about that already, though.

**Comment by vicary** [comment]
@dmoagx I cannot locate a similar issue, but feel free to close this if that is the case.

**Comment by Haravikk** [comment]
This sounds to me like a misuse of the BINARY field; if you're entering a hex string into the text field editor for a BINARY field then you're not actually storing the value you expect at all.

To clarify; a hex string is a sequence of characters representing octet (4-bit) values, but if you enter this as text then you're entering a series of regular text characters, in which case each one will occupy 8-bits (or more, depending upon encoding), this is why your input appears to be truncated those 34 characters are actually 68 bytes of data, which won't fit. The text field editor is not interpreting your input as a hex sequence, it's just entering what you give it as if it were text.

You can prove this to yourself by turning show as hex on in the view menu; you'll see that the value you stored isn't what you see in the pop-over editor at all.


The correct way to enter hex values in Sequel Pro right now seems to be with a query, like so:

```
INSERT INTO `foo` SET `uuid` = X'B4AF0D34726311E5A4E20242AC110005';
```

Plus whatever else your table requires. Instead of the X'blah' notation you can also use UNHEX('blah') for the same result, and when running your queries you should use `SELECT HEX(uuid)` to get the hex sequence out.

**Comment by vicary** [comment]
That's what I have been mentioning in this issue, TBH I don't care amongst `UNHEX()`, `X'...'` or even plain scribbles.

I am seeking consistency here, since the pop-up sheet translates it into `0x...`, I assume I can edit and stores new (or old) values back into the field with the same format beginning with `0x`. The same applies to inline editing.

But both of them has their own issues.
